src/geo/jts.clj

```diff
@@ -303,43 +283,90 @@
     (isGeometryChanged [_]
       true)))
 
+(defn- geom-srid?
+  "Check if a Geometry has a valid SRID."
+  [g]
+  (let [geom-srid (crs/get-srid g)]
+    (and (not= 0 geom-srid)
+         (not (nil? geom-srid)))))
+
 (defn- tf
   "Transform a Geometry by applying CoordinateTransform to the Geometry.
   When the target CRS has an SRID, set the geometry's SRID to that."
-  [^Geometry g ^CoordinateTransform transform]
+  [^Geometry g ^CoordinateTransform transform ^GeometryFactory gf]
   (let [g (.copy g)]
     (.apply g (transform-coord-seq-filter transform))
-    (set-srid g (crs/get-srid (crs/get-target-crs transform)))))
+    (set-srid g gf)))
 
 (defn transform-geom
   "Transform a Geometry.
-  When a single CoordinateTransform is passed, apply that transform to the Geometry. When the target CRS
-  has an SRID, set the geometry's SRID to that. When a single Transformable target is passed, attempt to
-  find the geometry's CRS to generate and apply a CoordinateTransform. When two CRSs are passed as
-  arguments, generate a CoordinateTransform and apply accordingly."
+  When a Geometry and CoordinateTransform is passed, apply that transform to the Geometry.
```

Comment:
Sorry @willcohen still being slow to review this in bits and pieces. The PR mostly looks good to me, but this function gives me a little pause. Specifically there are just a ton of cases here and I'm wondering if there's any way we can get out of supporting all of them in one function.

Is there one of these that you think is the "best" / sanctioned way to achieve what the function does? If so, could we strip this down to handling 1 or 2 of them and if needed give a separate function that can help people pipeline the process of acquiring the proper arguments to call the sanctioned one if needed.

I'm sure we can add enough tests to cover all of these branches, but I worry it will be hard to know which ones are actually used by people, so we'll be stuck supporting them forever without having a great insight into which branches matter.

I'm afraid I haven't worked with all of the geom transformation stuff that much myself, so I'm relying largely on your feedback as a user here, but what do you think of that? Is there a way to strip this down some?

Reply:
It's a good point for sure, and the conditional is definitely less than optimal. As I see it, there's actually one 'optimized' single base case, which is what goes to `tf` under this new functionality -- pass in both a Transform and a Factory. All of the conditional logic is to add that single arity while preserving the existing behavior before the Factory was ever an option.

In 2.0, there was only `(transform-geom g source target)`.
In 3.0, there was an additional `(transform-geom g transform)`, which the 2.0 arity now boils down to.
Effectively here, I'm trying to add a `(transform-geom g transform factory)`, which both the 2.0 and 3.0 versions could boil down to, except since we're only dispatching on arities and not types it becomes hard to do without conditionally figuring out what's a factory, what's a transform, what's a source, and what's a target.

There is another option -- we could use, yet again, a Protocol to try to address this. If `Transformable` also added `create-transform` as one of its defined functions and included `CoordinateTransform`, I think that most of this logic could pass back and forth in there. 

From there, we have two options. First, `transform-geom` could still retain the proposed `[g t], [g c1 c2], [g c1 c2 factory]` arity options, but its internal logic would be simpler. Second, `transform-geom` could be left largely as is for 3.1 and makes sense for basic cases, but which would internally defer to some new `transform-geom-factory` function that takes those additional external factories as a final argument each time.

I'm sure the details will make this a little more complicated, so I'll try out something along these lines in the next few days and see if something makes sense.